x/stake/handler.go

```diff
@@ -0,0 +1,318 @@
+	return sdk.Result{}
+}
+
+func (k Keeper) handleMsgDelegate(ctx sdk.Context, msg MsgDelegate) sdk.Result {
+
+	candidate, found := k.GetCandidate(ctx, msg.CandidateAddr)
+	if !found {
+		return ErrBadCandidateAddr().Result()
+	}
+	if msg.Bond.Denom != k.GetParams(ctx).BondDenom {
+		return ErrBadBondingDenom().Result()
+	}
+	if ctx.IsCheckTx() {
+		return sdk.Result{
+			GasUsed: GasDelegate,
+		}
+	}
+	return k.delegateWithCandidate(ctx, msg.DelegatorAddr, msg.Bond, candidate).Result()
+}
+
+func (k Keeper) delegateWithCandidate(ctx sdk.Context, delegatorAddr sdk.Address,
+	bondAmt sdk.Coin, candidate Candidate) sdk.Error {
+
+	if candidate.Status == Revoked { //candidate has been withdrawn
+		return ErrBondNotNominated()
+	}
+
+	// Get or create the delegator bond
+	existingBond, found := k.getDelegatorBond(ctx, delegatorAddr, candidate.Address)
+	if !found {
+		existingBond = DelegatorBond{
+			DelegatorAddr: delegatorAddr,
+			CandidateAddr: candidate.Address,
+			Shares:        sdk.ZeroRat,
+		}
+	}
+
+	// Account new shares, save
+	err := k.BondCoins(ctx, existingBond, candidate, bondAmt)
+	if err != nil {
+		return err
+	}
+	k.setDelegatorBond(ctx, existingBond)
+	k.setCandidate(ctx, candidate)
+	return nil
+}
+
+// Perform all the actions required to bond tokens to a delegator bond from their account
+func (k Keeper) BondCoins(ctx sdk.Context, bond DelegatorBond, candidate Candidate, amount sdk.Coin) sdk.Error {
+
+	_, err := k.coinKeeper.SubtractCoins(ctx, bond.DelegatorAddr, sdk.Coins{amount})
+	if err != nil {
+		return err
+	}
+	newShares := k.candidateAddTokens(ctx, candidate, amount.Amount)
+	bond.Shares = bond.Shares.Add(newShares)
+	k.setDelegatorBond(ctx, bond)
+	return nil
+}
+
+func (k Keeper) handleMsgUnbond(ctx sdk.Context, msg MsgUnbond) sdk.Result {
+
+	// check if bond has any shares in it unbond
+	bond, found := k.getDelegatorBond(ctx, msg.DelegatorAddr, msg.CandidateAddr)
+	if !found {
+		return ErrNoDelegatorForAddress().Result()
+	}
+	if !bond.Shares.GT(sdk.ZeroRat) { // bond shares < msg shares
+		return ErrInsufficientFunds().Result()
+	}
+
+	// if shares set to special case Max then we're good
+	if msg.Shares != "MAX" {
+		// test getting rational number from decimal provided
+		shares, err := sdk.NewRatFromDecimal(msg.Shares)
+		if err != nil {
+			return err.Result()
+		}
+
+		// test that there are enough shares to unbond
+		if !bond.Shares.GT(shares) {
+			return ErrNotEnoughBondShares(msg.Shares).Result()
+		}
+	}
+	if ctx.IsCheckTx() {
+		return sdk.Result{
+			GasUsed: GasUnbond,
+		}
+	}
+
+	// retrieve the amount of bonds to remove (TODO remove redundancy already serialized)
+	var shares sdk.Rat
+	var err sdk.Error
+	if msg.Shares == "MAX" {
+		shares = bond.Shares
+	} else {
+		shares, err = sdk.NewRatFromDecimal(msg.Shares)
+		if err != nil {
+			return err.Result()
+		}
+	}
+
+	// subtract bond tokens from delegator bond
+	if bond.Shares.LT(shares) { // bond shares < msg shares
```

Comment:
Redundant conditional? `shares` should already be either "MAX" (== `bond.Shares`) or less than `bond.Shares` as tested above.